File: discrete-math/lab1/F.java
import java.util.*;
 
public class F {
    public static int[] OneForAll(int n, int k, int[][] x) {
        int[] one = new int[k];
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < n; j++) {
                if (x[i][j] != -1) {
                    one[i] += 1;
                }
            }
        }
 
        int[] bruh = new int[n + 1];
        for (int i = 0; i < n; i++) {
            bruh[i] = -1;
        }
        int key = 0;
        for (int i = 0; i < k; i++) {
            if (one[i] == 1) {
                for (int j = 0; j < n; j++) {
                    if (x[i][j] != -1) {
                        if ((bruh[j] != -1) && (bruh[j] != x[i][j])) {
                            key = 1;
                        } else {
                            bruh[j] = x[i][j];
                        }
                    }
                }
            }
        }
        int count = 0;
        for (int i = 0; i < k; i++) {
            if (one[i] > 1) {
                count = 0;
                for (int j = 0; j < n; j++) {
                    if (((x[i][j] == 0) && (bruh[j] == 1)) || ((x[i][j] == 1) && (bruh[j] == 0)) || (x[i][j] == -1)) {
                        count += 1;
                    }
                }
                //System.out.println("count: " + count);
                if (count == n) {
                    key = 1;
                    break;
                }
            }
        }
        bruh[n] = key;
        return bruh;
 
    }
    public static int[][] putZero(int n, int k, int[][] x, int pp) { // x_k элемент приравнивается нулю
        for (int i = 0; i < k; i++) {
            if (x[i][pp] == 0) {
                for (int j = 0; j < n; j++) {
                    x[i][j] = -1;
                }
            } 
            if (x[i][pp] == 1) {
                x[i][pp] = -1;
            }
        }
        return x;
    } 
    public static int[][] putOne(int n, int k, int[][] x, int pp) { // x_k элемент приравнивается единице
        for (int i = 0; i < k; i++) {
            if (x[i][pp] == 1) {
                for (int j = 0; j < n; j++) {
                    x[i][j] = -1;
                }
            } 
            if (x[i][pp] == 0) {
                x[i][pp] = -1;
            }
        }
        return x;
    } 
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int n = input.nextInt(); // Количество столбцов
        int k = input.nextInt(); // Количество строк
        int [][] x = new int[100][100];
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < n; j++) {
                x[i][j] = input.nextInt();
            }
        }
        int count;
        int[] mas;
        input.close();
        while (true) {
            mas = OneForAll(n, k, x);
            System.err.println("OneForAll: " + Arrays.toString(mas));
            if (mas[n] == 1) {
                System.out.println("YES");
                break;
            }
            count = 0;
            for (int i = 0; i < n; i++) {
                if (mas[i] != -1) {
                    count += 1;
                }
            }
            if (count == 0) {
                System.out.println("NO");
                break;
            }
 
            for (int i = 0; i < n; i++) {
                if (mas[i] == 0) {
                    x = putZero(n, k, x, i);
                }
                if (mas[i] == 1) {
                    x = putOne(n, k, x, i);
                }
            }
             
            System.err.println("------");
            for (int i = 0; i < k; i ++) {
                for (int j = 0; j < n; j++) {
                    System.err.print(x[i][j] + " ");
                }
                System.err.println();
            }
             
        }
        //System.out.println(Arrays.toString(OneForAll(n, k, x)));
    }
}

File: discrete-math/lab1/F.py
def OneForAll(n, k, x):
    one = [0]*k
    for i in range(k):
        for j in range(n):
            if x[i][j] != -1:
                one[i] += 1
    bruh = [-1]*(n+1)
    key = 0
    for i in range(k):
        if one[i] == 1:
            for j in range(n):
                if x[i][j] != -1:
                    if ((bruh[j] != -1) and (bruh[j] != x[i][j])):
                        key = 1
                    else:
                        bruh[j] = x[i][j]
    #print(key)
    count = 0
    for i in range(k):
        if one[i] > 1:
            count = 0
            for j in range(n):
                if (((x[i][j] == 0) and (bruh[j] == 1)) or ((x[i][j] == 1) and (bruh[j] == 0)) or (x[i][j] == -1)):
                    #((bruh[j] != -1) and (bruh[j] != x[i][j])):
                    count += 1
            if count == n:
                key = 1
                break
    bruh[n] = key
    return bruh
def putOne(n, k, x, pp):
    for i in range(k):
        if x[i][pp] == 1:
            for j in range(n):
                x[i][j] = -1
        if x[i][pp] == 0:
            x[i][pp] = -1
    return x
def putZero(n, k, x, pp):
    for i in range(k):
        if x[i][pp] == 0:
            for j in range(n):
                x[i][j] = -1
        if x[i][pp] == 1:
            x[i][pp] = -1
    return x
n, k = map(int, input().split())
x = [[] for j in range(100)]
for i in range(k):
    x[i] = list(map(int, input().split()))
count = 0
while True:
    mas = OneForAll(n, k, x)
    #print(mas)
    if mas[n] == 1:
        print("YES")
        break
    count = 0
    for i in range(n):
        if mas[i] != -1:
            count += 1
    if count == 0:
        print("NO")
        break
    for i in range(n):
        if mas[i] == 0:
            x = putZero(n, k, x, i)
        if mas[i] == 1:
            x = putOne(n, k, x, i)
    #print("-------")
    #for i in range(k):
        #print(*x[i])
